rackcity/utils/change_planner_utils.py

```diff
@@ -1,54 +1,153 @@
     filtered_assetsCP, filter_failure_response = get_filtered_query(
         assetsCP,
+        data,
+    )
+    if filter_failure_response:
+        return (None, None, filter_failure_response)
+    return (filtered_assets, filtered_assetsCP, None)
+
+
+def sort_serialized_assets(all_assets, data):
+    sorted_assets = all_assets
+    if 'sort_by' in data:
+        for sort in data['sort_by']:
+            if (
+                'field' in sort
+                and 'ascending' in sort
+                and isinstance(sort['field'], str)
+                and isinstance(sort['ascending'], bool)
+            ):
+                sorted_assets.sort(
+                    key=lambda i: i[sort['field']],
+                    reverse=not sort['ascending'],
+                )
+    return sorted_assets
+
+
+def get_page_of_serialized_assets(all_assets, query_params):
+    page = int(query_params.get('page'))
+    page_size = int(query_params.get('page_size'))
+    start = (page - 1) * page_size
+    end = page * page_size
+    if start > len(all_assets):
+        raise Exception("Page " + str(start) + " does not exist")
+    elif end > len(all_assets):
+        return all_assets[start:]
+    else:
+        return all_assets[start:end]
+
+
+def get_many_assets_response_for_cp(request, change_plan):
+    should_paginate = should_paginate_query(request.query_params)
+    if should_paginate:
+        page_failure_response = get_invalid_paginated_request_response(
+            request.query_params
+        )
+        if page_failure_response:
+            return page_failure_response
+
+    assets, assetsCP, filter_failure_response = get_filtered_assets_for_cp(
+        change_plan,
         request.data,
     )
     if filter_failure_response:
         return filter_failure_response
     asset_serializer = RecursiveAssetSerializer(
-        filtered_assets,
+        assets,
         many=True,
     )
     assetCP_serializer = RecursiveAssetCPSerializer(
-        filtered_assetsCP,
+        assetsCP,
         many=True,
     )
     all_assets = asset_serializer.data + assetCP_serializer.data
+
+    sorted_assets = sort_serialized_assets(all_assets, request.data)
+
+    if should_paginate:
+        try:
+            assets_data = get_page_of_serialized_assets(
+                sorted_assets,
+                request.query_params,
+            )
+        except Exception as error:
+            return JsonResponse(
+                {
+                    "failure_message":
+                        Status.ERROR.value + GenericFailure.PAGE_ERROR.value,
+                        "errors": str(error)
+                },
+                status=HTTPStatus.BAD_REQUEST,
+            )
+    else:
+        assets_data = sorted_assets
+
     return JsonResponse(
-        {"assets": all_assets},
+        {"assets": assets_data},
         status=HTTPStatus.OK,
     )
+
+
+def get_page_count_response_for_cp(request, change_plan):
```

Comment:
Updated /assets/pages endpoint to support change plans and filtering on change plans